src/test/regress/sql/local_shard_execution_dropped_column.sql

```diff
@@ -0,0 +1,58 @@
+CREATE SCHEMA local_shard_execution_dropped_column;
+SET search_path TO local_shard_execution_dropped_column;
+
+SET citus.next_shard_id TO 2460000;
+
+-- the scenario is described on https://github.com/citusdata/citus/issues/5038
+
+-- first stop the metadata syncing to the node do that drop column
+-- is not propogated
+SELECT stop_metadata_sync_to_node('localhost',:worker_1_port);
+SELECT stop_metadata_sync_to_node('localhost',:worker_2_port);
+
+-- create a distributed table, drop a column and sync the metadata
+SET citus.shard_replication_factor TO 1;
+CREATE TABLE t1 (a int, b int, c int UNIQUE);
+SELECT create_distributed_table('t1', 'c');
+ALTER TABLE t1 DROP COLUMN b;
+SELECT start_metadata_sync_to_node('localhost',:worker_1_port);
+SELECT start_metadata_sync_to_node('localhost',:worker_2_port);
```

Comment:
Another case that I think would result in a similar bug is:
```
-- start metadata sync now
CREATE TABLE t2 (a int, b int, c int UNIQUE);
ALTER TABLE t2 DROP COLUMN b;
select citus_move_shard_placement(...); -- now the newly created shard has different attribute numbers
-- run multi shard query on node where the shard was moved to
```

As far as I understand the current fix only reparses the query once for a random shard, and uses those attribute numbers for each shard. If that's indeed the case, then I think this will still fail with the current fix, because the attribute numbers are not the same for all shards.

Reply:
> -- run multi shard query on node where the shard was moved to

The bug is only relevant for single shard queries that can have deferredPruning (e.g., INSERTs and fast-path queries). Because we do caching only for these types of queries


> As far as I understand the current fix only reparses the query once for a random shard,

The caching mechanism in general does the caching per shard/per query, see `GetCachedLocalPlan()` for the details.

> select citus_move_shard_placement(...); -- now the newly created shard has different attribute numbers

In that case, the shell table would have the dropped column and not the shard -- as the shard is re-created. 

```SQL
select attrelid::regclass, attname, attnum, attisdropped from pg_attribute WHERE attrelid IN ('t2'::regclass, 't2_102106'::regclass) and attname NOT IN ('tableoid','cmax', 'xmax', 'cmin', 'xmin', 'ctid') ;
 attrelid  |           attname            | attnum | attisdropped 
-----------+------------------------------+--------+--------------
 t2        | ........pg.dropped.1........ |      1 | t
 t2        | c                            |      2 | f
 t2_102106 | c                            |      1 | f
(3 rows)
```

However, because the patch deparses the query, we actually do not care about the discrepancy between attribute numbers anymore. Because, now Postgres will give us the accurate attribute numbers anyway.

But, I think still this is a good test to have -- even both directions of moves--, thanks for pointing out.

Reply:
> However, because the patch deparses the query, we actually do not care about the discrepancy between attribute numbers anymore. Because, now Postgres will give us the accurate attribute numbers anyway.

I'm actually not sure about this, will look more. There could be some issues, but somehow it works fine

Reply:
Ok, I understand now why that is not becoming an issue.

To start with, the situation you described is applicable not only for local cached plans but also any remote query as well where the table attribute differs from the shard attribute.

The deparsing logic roughly works as follow:
- We get a query from the user, and put it as `jobQuery`
- For each task, we call `UpdateRelationToShardNames(jobquery, relationShardList)`, which does a trick where the relations are replaces with functions, see :

```C
/*
 * SetRangeTblExtraData adds additional data to a RTE, overwriting previous
 * values, if present.
 *
 * The data is stored as RTE_FUNCTION type RTE of a special
 * citus_extradata_container function, with the extra data serialized into the
 * function arguments. That works, because these RTEs aren't used by Postgres
 * to any significant degree, and Citus' variant of ruleutils.c knows how to
 * deal with these extended RTEs. Note that rte->eref needs to be set prior
 * to calling SetRangeTblExtraData to ensure the funccolcount can be set
 * correctly.
 *
 * NB: If used for postgres defined RTEKinds, fields specific to that RTEKind
 * will not be handled by out/readfuncs.c. For the current uses that's ok.
 */
```
- And the deparsing logic replaces the table names with shard names

The short of the story is that we do the deparsing based on the `jobQuery`, meaning that we don't care about the shard table's attributes. But when we parse back the queryString that is on `shard table`, Postgres would already give us the accurate attribute number for the shard.

Does that make sense to you @JelteF ?

Reply:
We cache a separate plan for every anchor shard (see GetCachedLocalPlan), and therefore would deparse+parse seperately for every shard. 

To do deparsing for a specific shard, we replace table RTEs with function calls that contain metadata about the shard in UpdateRelationToShardNames. We also modified the ruleutils logic to recognize those RTEs (CITUS_RTE_SHARD) and generate shard names. The reason we do this elaborate trick is that ruleutils normally only knows how to deparse table names that exist in pg_class.

The local Citus plans are cached within the cached prepared statement plan. Invasive maneuvers like pg_dist_placement updates (read: shard moves) invalidate the prepared statement cache by sending a invalidation for the relation (e.g. see citus_dist_placement_cache_invalidate, or InsertShardPlacementRow) when they commit, which implicitly invalidates the cached Citus plans as well.

Reply:
Thanks for all the details both, makes sense that this fixes that too then. But like onder said, might still be nice to add it as a test though.

Reply:
> Thanks for all the details both, makes sense that this fixes that too then. But like onder said, might still be nice to add it as a test though.

Yes, makes sense added such an example